### PR TITLE
fix(python): group_by.map_groups works with expressions in group_by

### PR DESCRIPTION
**Pull request by deanm0000**
fixes https://github.com/pola-rs/polars/issues/13442

### CODE REVIEW

File: py-polars/polars/lazyframe/frame.py [hunk 1]
```diff
@@ -3223,7 +3223,8 @@ def group_by(
         """
         exprs = parse_as_list_of_expressions(*by, **named_by)
         lgb = self._ldf.group_by(exprs, maintain_order)
-        return LazyGroupBy(lgb)
+        lf = self
+        return LazyGroupBy(lgb, *by, _lf=lf, maintain_order=maintain_order, **named_by)
```
Comment (stinodego) [inline py-polars/polars/lazyframe/frame.py:3227]: Why did you change the LazyGroupBy? Seems unrelated and unnecessary.

File: py-polars/polars/dataframe/group_by.py [hunk 2]
```diff
@@ -311,16 +311,12 @@ def map_groups(self, function: Callable[[DataFrame], DataFrame]) -> DataFrame:
...
-
-        return self.df.__class__._from_pydf(
-            self.df._df.group_by_map_groups(
-                list(self.by), function, self.maintain_order
+        self.by_str_cols = self.df.lazy().select(*self.by, **self.named_by).columns
+        df = self.df.with_columns(*self.by, **self.named_by)
+
+        return df.__class__._from_pydf(
+            df._df.group_by_map_groups(
+                list(self.by_str_cols), function, self.maintain_order
```
Comment (stinodego) [inline py-polars/polars/dataframe/group_by.py:319]: The `list` call here is unnecessary.

File: py-polars/polars/dataframe/group_by.py [hunk 3]
```diff
@@ -311,16 +311,12 @@ def map_groups(self, function: Callable[[DataFrame], DataFrame]) -> DataFrame:
...
-            raise TypeError(msg)
-        if not all(isinstance(c, str) for c in self.by):
-            msg = "cannot call `map_groups` when grouping by an expression"
-            raise TypeError(msg)
-
-        return self.df.__class__._from_pydf(
-            self.df._df.group_by_map_groups(
-                list(self.by), function, self.maintain_order
+        self.by_str_cols = self.df.lazy().select(*self.by, **self.named_by).columns
+        df = self.df.with_columns(*self.by, **self.named_by)
```
Comment (stinodego) [inline py-polars/polars/dataframe/group_by.py:315]: I think this might handle name clashes incorrectly, but I'm not sure. Could you add some tests for this?

**Review by stinodego — CHANGES_REQUESTED** [review]
Thanks @deanm0000 , a few comments!

File: py-polars/polars/lazyframe/frame.py (hunk 1 above)
Comment (deanm0000) [reply py-polars/polars/lazyframe/frame.py:3227]: In the def of the lazy version of map_groups I need to start from the lf that preceded the call to `group_by` (just like in the eager one) but b/c in the lazy version it turns it into a new object in the frame context I don't have it in the LazyGroupBy context so I had to pass it through. Is there a better way to accomplish that?

Comment (stinodego) [reply py-polars/polars/lazyframe/frame.py:3227]: The lazy version of `map_groups` already worked correctly, right? It was just the eager version that did not support expressions?

Comment (deanm0000) [reply py-polars/polars/lazyframe/frame.py:3227]: the lazy version doesn't raise but it also doesn't respect expressions other than taking the names. There's an example in the 2nd code block here https://github.com/pola-rs/polars/issues/13442

File: py-polars/polars/dataframe/group_by.py (hunk 2 above)
Comment (deanm0000) [reply py-polars/polars/dataframe/group_by.py:319]: will remove, not sure why I put that there.

File: py-polars/polars/dataframe/group_by.py (hunk 3 above)
Comment (deanm0000) [reply py-polars/polars/dataframe/group_by.py:315]: I don't think it executes twice. In line 314 it uses a lazy select so that it doesn't materialize anything except a new schema from which I take the columns as a list of column names. In line 315, it does eagerly materialize those expressions into columns before the `group_by_map_groups`.

**Comment by mcrumiller** [comment]
Why was this closed? AFAIK the issue is still open.

**Comment by deanm0000** [comment]
@mcrumiller hmm I don't know/remember. Probably closed this one from mobile by accident instead of something else.
